cmd/nodecmd/join.go

```diff
@@ -0,0 +1,478 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package nodecmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/units"
+
+	"github.com/ava-labs/avalanche-cli/pkg/ansible"
+
+	"github.com/ava-labs/avalanchego/vms/platformvm"
+
+	subnetcmd "github.com/ava-labs/avalanche-cli/cmd/subnetcmd"
+	"github.com/ava-labs/avalanche-cli/pkg/constants"
+	"github.com/ava-labs/avalanche-cli/pkg/models"
+	"github.com/ava-labs/avalanche-cli/pkg/prompts"
+	"github.com/ava-labs/avalanche-cli/pkg/subnet"
+	"github.com/ava-labs/avalanche-cli/pkg/ux"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/spf13/cobra"
+)
+
+var (
+	deployTestnet   bool
+	deployMainnet   bool
+	keyName         string
+	subnetName      string
+	useLedger       bool
+	ledgerAddresses []string
+	weight          uint64
+	duration        time.Duration
+
+	ErrMutuallyExlusiveKeyLedger = errors.New("--key and --ledger,--ledger-addrs are mutually exclusive")
+	ErrStoredKeyOnMainnet        = errors.New("--key is not available for mainnet operations")
+	ErrNoBlockchainID            = errors.New("failed to find the blockchain ID for this subnet, has it been deployed/created on this network?")
+)
+
+func newJoinCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "join [subnetName]",
+		Short: "(ALPHA Warning) Join a subnet as a validator",
+		Long: `(ALPHA Warning) This command is currently in experimental mode.
+
+The node join command enables a Primary Network Validator to also be a validator
+of a Subnet. If The command is run before the node is bootstrapped on the Primary Network, 
+the command will fail. You can check the bootstrap status by calling 
+avalanche node status`,
+		SilenceUsage: true,
+		Args:         cobra.ExactArgs(1),
+		RunE:         joinSubnet,
+	}
+	cmd.Flags().StringVar(&subnetName, "subnet", "", "specify the subnet the node is validating")
+
+	return cmd
+}
+
+func parseBootstrappedOutput(filePath string) (bool, error) {
+	jsonFile, err := os.Open(filePath)
+	if err != nil {
+		return false, err
+	}
+	defer jsonFile.Close()
+	byteValue, _ := io.ReadAll(jsonFile)
+	var result map[string]interface{}
+	err = json.Unmarshal(byteValue, &result)
+	if err != nil {
+		return false, err
+	}
+	isBootstrappedInterface, ok := result["result"].(map[string]interface{})
+	if ok {
+		isBootstrapped, ok := isBootstrappedInterface["isBootstrapped"].(bool)
+		if ok {
+			return isBootstrapped, nil
+		}
+	}
+	return false, nil
+}
+
+func parseSubnetSyncOutput(filePath string) (string, error) {
+	jsonFile, err := os.Open(filePath)
```

Comment:
maybe we can have one function for the block

Reply:
I believe the code block is almost repeated for the three outputs, json, result, only change the field name and in one case it is a bool